### DISCUSSION TITLE
How does MediaCMS encoding works when using multiple servers?

### DISCUSSION THREAD

**Discussion by ricardoteix** [mediacms-io/mediacms#851] [General]
Hello everyone. 
I am using mediacms on AWS. My infrastructure uses a RDS database and 2 EC2 Ubuntu instances as servers, lets call they as server A and server B.  The media_files folder are shared between these 2 server using s3fs.

The idea of server B is to receive the specific traffic to upload and media enconding, while server A does the rest of the job.
So my LoadBalancer is configured to send all POST traffic coming from the routes **/fu/upload/*, /fu/upload, /fu/upload/, /upload** to server B, and all other to server A.
When I access by browser it seems to be ok. The upload was going as expected, but when I see the realtime CPU usage afeter the upload both of them are in 100%. It is seems both are encoding the video and this is the focus of my doubt. 

In this case is really both processing as a shared task and speeding up the encoding process or they are doing the same job and wasting resources?
And it there a way to make only server B to encode the video and let server A alone to do other jobs?

Thanks to all.

**Comment by mgogoulos** [comment]
Hi, I'd like to see more details about this setup, as I have not used it myself (I have mostly been hosting everything in a single server). 
How does MediaCMS run on server A and server B? Is it through docker-compose or through single server install, in both servers? 
The process celery_long is responsible for picking the tasks and performing the transcoding, so you can see that if this is running in server A and server B, and assuming there is a single redis that they both listen to (and is holding the job information), both servers are picking jobs. The solution here would be to run celery_long on server B only, and stop it on server A. 
As far as the uploading is related, it is handled by the "web" task, so mediacms process on single server install, or web container on docker-compose. 

Hope this helps

**Reply by ricardoteix** [reply]
Thank you for the answer @mgogoulos.

I am using Terraform to get it all up and running. I adapted the install.sh and used it as my EC2 instance user-data. 
The Terraform also configure a AutoScaling to scale up and out the number of instances from 1 to 5 according to average CPU use. The 1 to 5 instances works like the server A. The server B was not part of autoscaling.
 
Some configurations in settings.py are made in this script too, including the use of Parameter Store, created via Terraform with all credentials to redis, postgres and some other credentials from it using  boto3.

The servers uses the same redis too, only one node. I tried a AWS Elasticache Redis cluster but an error occurred.

If you wanto to see all the IaC code look at this repository [https://github.com/ricardoteix/ct-oficinadeprojetos-12](https://github.com/ricardoteix/ct-oficinadeprojetos-12).

I will make some changes based on you pointed and let you know if it works.

Thank you again.

**Reply by ricardoteix** [reply]
Hello again, @mgogoulos .

Stopping the celety_long solved my problem. I was abble to add a tag on the EC2 to start the celery only in that instance I want to process the video. 

Thank you for the tip. If you wish to know more about my Terraform feel free to ask.

**Reply by ricardoteix** [reply]
Hi,

I've finally created an automation to start an EC2 instance specifically to process videos when a file arrives at the _/media_files/original/*_ folder on the S3 bucket. There are some issues, but for my goal, it was great.

I've created an English document to explain my project, and for those who want to know more, start from [here](https://github.com/ricardoteix/ct-oficinadeprojetos-12/blob/main/README_en.md).

Thank you again, @mgogoulos.

**Reply by mgogoulos** [reply]
Hi @ricardoteix , wondering how the setup is functioning, did you deploy your portal? Would love to read any insights. Also if you want to share it on https://github.com/mediacms-io/mediacms/discussions/categories/show-and-tell it would be great - if you can post a URL or screenshot, along with basic info about your use case and modifications you made ;)

**Reply by ricardoteix** [reply]
Hi @mgogoulos.

I've wrote some things about the project as a **show-and-tell** post [here](https://github.com/mediacms-io/mediacms/discussions/891).
Hope it helps to undertand some things I've made.
